workflow/controller/artifact_gc.go

```diff
@@ -71,6 +71,34 @@ func (woc *wfOperationCtx) garbageCollectArtifacts(ctx context.Context) error {
 	return nil
 }
 
+func (woc *wfOperationCtx) HasArtifactGC() bool {
+	// ArtifactGC can be defined on the Workflow level or on the Artifact level
+	// It may be defined in the Workflow itself or in a WorkflowTemplate referenced by the Workflow
+
+	// woc.execWf.Spec.Templates includes templates described directly in the Workflow as well as templates
+	// in a WorkflowTemplate that the entire Workflow is based on
+	for _, template := range woc.execWf.Spec.Templates {
+		for _, artifact := range template.Outputs.Artifacts {
+			strategy := woc.execWf.GetArtifactGCStrategy(&artifact)
+			if strategy != wfv1.ArtifactGCStrategyUndefined && strategy != wfv1.ArtifactGCNever {
+				return true
+			}
+		}
+	}
+
+	// need to go to woc.wf.Status.StoredTemplates in the case of a Step referencing a WorkflowTemplate
+	for _, template := range woc.wf.Status.StoredTemplates {
+		for _, artifact := range template.Outputs.Artifacts {
+			strategy := woc.execWf.GetArtifactGCStrategy(&artifact)
+			if strategy != wfv1.ArtifactGCStrategyUndefined && strategy != wfv1.ArtifactGCNever {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
```

Comment:
function moved from workflow_types.go to here where we can access both execWf for Spec and wf for Status